### PR TITLE
Inliner rewrite

### PR DESCRIPTION
**Pull request by sgpthomas**

### CODE REVIEW

File: calyx/src/ir/structure.rs [hunk 1]
```diff
@@ -39,6 +41,23 @@ impl Port {
...
+                cell.upgrade().unwrap().borrow().name.clone()
+            }
+            PortParent::Group(group) => {
+                group.upgrade().unwrap().borrow().name.clone()
+            }
+        }
+    }
+
+    /// Gets a key usable in a hashmap from a port
+    pub fn key(&self) -> (Id, Id) {
```
Comment (rachitnigam) [inline calyx/src/ir/structure.rs:58]: This seems far too specific to the HashMap used in the inliner.

File: calyx/src/ir/structure.rs [hunk 2]
```diff
@@ -15,21 +15,23 @@ pub enum Direction {
...
+#[derive(Debug, Clone)]
 pub struct Port {
     /// Name of the port
     pub name: Id,
     /// Width of the port
     pub width: u64,
     /// Direction of the port
     pub direction: Direction,
+    /// Name of parent
+    pub parent_name: Id,
```
Comment (rachitnigam) [inline calyx/src/ir/structure.rs:34]: I remain unconvinced that this is needed. Whenever you're ready, do mention where it is needed in the writeup for the PR.

File: calyx/src/ir/structure.rs (hunk 1 above)
Comment (sgpthomas) [reply calyx/src/ir/structure.rs:58]: I can move this for now, but I thought it might be useful in general for using Ports in hashmaps. This seems potentially useful for other anlayses.

File: calyx/src/ir/structure.rs (hunk 2 above)
Comment (sgpthomas) [reply calyx/src/ir/structure.rs:34]: Otherwise there is no way to mutate assignments based on whether it uses a hole from a particular group. This is because ports themselves are not unique and if you are mutating the group, then you can't look up the parent of a hole, because the group is already borrowed. You can't do this in two steps either, because there is no way to relate the immutable information gathering pass to the mutable assignment changing pass because again, ports are not unique. 

If you are really opposed to adding `parent_name`, then we can make ports unique in some other way. But that seems like a more complicated solution than is called for

Comment (rachitnigam) [reply calyx/src/ir/structure.rs:34]: Can you point out the place in the inliner where this logic applies?

My problem with exposing the `parent_name` is that we're back to square one: what is the "canonical" way of getting a parent? Do I use the parent pointer or do I get the parent name and call one of the `find_` methods on the components.

The principles of the IR is to have one way to do things and to not duplicate information. This violates both.

File: calyx/src/ir/analysis/graph.rs [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+                    Rc::clone(&self.graph[node_idx])
+                })
+            }
+            Direction::Output => unimplemented!(),
+        }
+    }
+
+    /// Restricts the graph of the analysis using a filter
+    /// function that specifies which edges to keep.
+    pub fn edge_induced_subgraph<F>(self, mut f: F) -> Self
```
Comment (rachitnigam) [inline calyx/src/ir/analysis/graph.rs:117]: This comment doesn't make sense w.r.t to the function name. What sub-graph is this "inducing"?

File: calyx/src/ir/analysis/graph.rs [hunk 4]
```diff
@@ -0,0 +1,159 @@
+use crate::{
```
Comment (rachitnigam) [inline calyx/src/ir/analysis/graph.rs:1]: Module comment needed at the top of the file.

File: calyx/src/ir/analysis/graph.rs [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+/// as a graph which makes certain types of queries more efficient to answer.
+#[derive(Clone)]
+pub struct Analysis {
+    nodes: HashMap<(Id, Id), NodeIndex>,
+    graph: CellGraph,
+}
+
+impl Analysis {
+    /// Start an analysis from a component.
+    pub fn from(component: &Component) -> Self {
```
Comment (rachitnigam) [inline calyx/src/ir/analysis/graph.rs:35]: More descriptive comment about what kind of analysis this is.

File: calyx/src/ir/analysis/graph.rs [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        );
+        Self { graph, nodes }
+    }
+
+    /// Checks if there are sub-graphs in the analysis graph.
+    pub fn has_cycles(&self) -> bool {
+        algo::is_cyclic_directed(&self.graph)
+    }
+
+    // pub fn visualize(&self) -> String {
```
Comment (rachitnigam) [inline calyx/src/ir/analysis/graph.rs:141]: Delete this?

File: calyx/src/ir/builder.rs [hunk 7]
```diff
@@ -209,23 +210,23 @@ impl<'a> Builder<'a> {
         outputs: Vec<(ir::Id, u64)>,
     ) -> RRC<ir::Cell> {
         let cell = Rc::new(RefCell::new(ir::Cell {
-            name,
+            name: name.clone(),
```
Comment (rachitnigam) [inline calyx/src/ir/builder.rs:213]: Unnecessary clone

File: calyx/src/ir/builder.rs [hunk 8]
```diff
@@ -209,23 +210,23 @@ impl<'a> Builder<'a> {
...
             ports: vec![],
             prototype: typ,
         }));
         // Construct ports
-        for (name, width) in inputs {
+        for (port_name, width) in inputs {
```
Comment (rachitnigam) [inline calyx/src/ir/builder.rs:218]: Unnecessary change (this and below this)

File: calyx/src/ir/guard.rs [hunk 9]
```diff
@@ -52,13 +114,41 @@ impl Guard {
             Guard::Geq(_, _) => ">=".to_string(),
             Guard::Leq(_, _) => "<=".to_string(),
             Guard::Not(_) => "!".to_string(),
-            Guard::Port(_) => panic!("No operator string for Guard::Port"),
+            Guard::Port(_) | Guard::True => {
+                panic!("No operator string for Guard::Port")
```
Comment (rachitnigam) [inline calyx/src/ir/guard.rs:118]: Incorrect `panic!` message.

File: calyx/src/ir/guard.rs [hunk 10]
```diff
@@ -52,13 +114,41 @@ impl Guard {
...
     }
 
-    ////////////// Convinience constructors ///////////////////
-    pub fn and(self, other: Guard) -> Self {
-        Guard::And(vec![self, other])
+    pub fn and_vec(&self, guards: &mut Vec<Guard>) -> Self {
+        flatten_and(guards);
+        let mut new;
+        if let Guard::And(inner) = &self {
+            new = inner.clone();
```
Comment (rachitnigam) [inline calyx/src/ir/guard.rs:127]: These `clones` seem unnecessary. Either make the function take ownership of `self` or define it as an associated function.

File: calyx/src/ir/guard.rs [hunk 11]
```diff
@@ -52,13 +114,41 @@ impl Guard {
...
+        new.retain(|guard| {
+            if matches!(guard, Guard::True) {
+                return false;
+            }
+
+            if let Guard::Port(p) = guard {
+                if p.borrow().is_constant(1, 1) {
+                    return false;
+                }
+            }
```
Comment (rachitnigam) [inline calyx/src/ir/guard.rs:144]: Move this to the top, and `return p.borrow().is_constant(1, 1)` followed by just returning `matches!(guard, Guard::True)`.

File: calyx/src/ir/structure.rs [hunk 12]
```diff
@@ -39,6 +39,31 @@ impl Port {
     pub fn is_hole(&self) -> bool {
         matches!(&self.parent, PortParent::Group(_))
     }
+
+    pub fn is_constant(&self, val: u64, width: u64) -> bool {
```
Comment (rachitnigam) [inline calyx/src/ir/structure.rs:43]: Comment describing this function.

File: calyx/src/passes/inliner.rs [hunk 13]
```diff
@@ -1,13 +1,14 @@
-use crate::lang::component::Component;
-use crate::lang::{
-    ast, context::Context, structure, structure_builder, structure_iter,
+use crate::{
+    errors::Error,
+    frontend::library::ast::LibrarySignatures,
+    ir::{
+        analysis::Analysis,
+        traversal::{Action, Named, VisResult, Visitor},
+        Assignment, Builder, Component, Control, Guard, Port,
```
Comment (rachitnigam) [inline calyx/src/passes/inliner.rs:7]: Prefer importing `crate::ir` and using `ir::Assignment` etc. instead of direct names.

File: calyx/src/passes/inliner.rs [hunk 14]
```diff
@@ -22,139 +23,124 @@ impl Named for Inliner {
...
-    // flatten groups (maybe temporary)
-    for idx in st.edge_idx().detach() {
-        st.get_edge_mut(idx).group = None;
-    }
-    st.groups = HashMap::new();
-    st.groups
-        .insert(None, (HashMap::new(), st.edge_idx().collect()));
-}
+        // add top_level[go] = this.go
+        let go_asgn = Assignment {
```
Comment (rachitnigam) [inline calyx/src/passes/inliner.rs:62]: See the uses of the `build_assignments!` macro in `compile_control`

File: calyx/src/passes/inliner.rs [hunk 15]
```diff
@@ -22,139 +23,124 @@ impl Named for Inliner {
...
-        inline_hole(st, "go".to_string());
-        inline_hole(st, "done".to_string());
+        // add the continuous assignment edges
+        assignments
+            .append(&mut comp.continuous_assignments.drain(..).collect());
+
+        // remove edges that write to a hole
+        assignments.retain(|asgn| !asgn.dst.borrow().is_hole());
+
+        // move direct reads from holes into the guard so they can be inlined
```
Comment (rachitnigam) [inline calyx/src/passes/inliner.rs:114]: Not sure what this comment means. Maybe an example will help.

File: calyx/src/passes/inliner.rs [hunk 16]
```diff
@@ -22,139 +23,124 @@ impl Named for Inliner {
...
+        let mut builder = Builder::from(comp, sigs, false);
```
Comment (rachitnigam) [inline calyx/src/passes/inliner.rs:115]: Isn't there a builder already in scope?

File: calyx/src/passes/inliner.rs [hunk 17]
```diff
@@ -22,139 +23,124 @@ impl Named for Inliner {
...
+                        None
+                    }
+                })
+            });
+        }
+        comp.continuous_assignments = assignments;
 
-        comp.control = Control::empty();
+        // remove all groups
+        comp.groups = vec![];
```
Comment (rachitnigam) [inline calyx/src/passes/inliner.rs:141]: Use `Vec::clear()`

File: calyx/src/utils.rs [hunk 18]
```diff
@@ -91,6 +91,11 @@ impl OutputFile {
     }
 }
 
+pub trait Keyable {
```
Comment (rachitnigam) [inline calyx/src/utils.rs:94]: Description comment

File: calyx/src/ir/analysis/graph.rs (hunk 3 above)
Comment (sgpthomas) [reply calyx/src/ir/analysis/graph.rs:117]: It is just a subgraph that is created (induced) using a subset of edges. https://mathworld.wolfram.com/Edge-InducedSubgraph.html

File: calyx/src/passes/inliner.rs (hunk 16 above)
Comment (sgpthomas) [reply calyx/src/passes/inliner.rs:115]: builder requires mutable access to comp as long as it is alive which prevents other mutable borrows. Reconstructing it shortens the lifespan of the builder. I'll add some comments to make this more clear

Comment (rachitnigam) [reply calyx/src/passes/inliner.rs:115]: `builder.component` returns a mutable reference to the component. I've used this pattern once in `compile_control`. Maybe we can have a function that operates on the mutable component reference defined on the `builder`?

File: calyx/src/ir/analysis/graph.rs (hunk 3 above)
Comment (rachitnigam) [reply calyx/src/ir/analysis/graph.rs:117]: This makes sense only if you think of assignments as edges. The default representation says nothing about assignments being edges. The comment needs to explicate that.

File: calyx/src/ir/builder.rs [hunk 19]
```diff
@@ -67,9 +67,9 @@ impl<'a> Builder<'a> {
         }));
 
         // Add default holes to the group.
-        for (name, width) in vec![("go", 1), ("done", 1)] {
+        for (port_name, width) in vec![("go", 1), ("done", 1)] {
             let hole = Rc::new(RefCell::new(ir::Port {
-                name: name.into(),
+                name: port_name.into(),
```
Comment (rachitnigam) [inline calyx/src/ir/builder.rs:72]: Unnecessary change.

**Comment by rachitnigam** [comment]
One unnecessary change marked. Merge after fixing.
